Super Cache: add an admin notice to encourage migration to Jetpack Boost

donnchawp opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->
For site owners who are using the default settings of WP Super Cache they could use Jetpack Boost and see much the same caching results. This PR adds an admin notice with a "Migrate now" button that installs and activates Boost. It will set a flag that tells Boost to activate the Cache module.
This PR also adds a new function, `wpsc_is_boost_current()` that checks if the installed version of Boost is newer than or equal to 3.4.0. That fixes a different issue linked below.

Fixes #37743 #37932

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Add a function wpsc_jetpack_boost_notice() that displays the notice. It will only display on the WPSC settings page.
* Add a function wpsc_dismiss_boost_notice() that allows the notice to be hidden.
* Add a function wpsc_is_boost_compatible() that checks if WPSC is using only basic caching features.
* Add a function wpsc_is_boost_current() to check that Boost version is current.
* Add JavaScript that catches clicks on the dismiss button so that action is recorded by wpsc_dismiss_boost_notice()
* Style the admin notice
* Add a nonce for dismissing the admin notice.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
pc9hqz-2QH-p2

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->
* Make sure your WPSC install has a default setting. Delete wp-content/wp-cache-config.php and visit the settings page, enable caching from Easy page.
* Delete Jetpack Boost plugin from your site and then reload WPSC settings page to see the admin notice.
* Clicking on the "Migrate Now" button will download, install and activate Boost and take you to the Boost settings page.
* Go to Plugins page and deactivate Jetpack Boost.
* Reload the WPSC settings page, click on the "Migrate Now" button again and it should activate Boost and take you there.
* Reload the WPSC settings page and the admin notice should be gone now.
* Delete or deactivate Boost and reload WPSC settings page to show admin notice again.
* Click the Dismiss button. It disappears, reload the page and it should stay gone.
* Delete the "wpsc_dismissed_boost_admin_notice" user option to show the admin notice again.
* `delete_user_option( get_current_user_id(), 'wpsc_dismissed_boost_admin_notice' );`

donnchawp committed:
Add an admin notice to encourage migration to Jetpack Boost

donnchawp committed:
changelog

donnchawp committed:
Boost age check only applies if Boost is installed.

donnchawp committed:
Add boost error and spinner to admin notice

donnchawp committed:
Rename banner error to migration error.

donnchawp committed:
Update animations so they work on any migration button

pyronaur commented on projects/plugins/super-cache/inc/boost.php:
```diff
@@ -11,6 +11,60 @@ function wpsc_get_boost_migration_config() {
+	</div>
+	<?php
+}
+if ( isset( $_GET['page'] ) && $_GET['page'] === 'wpsupercache' ) { // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+	add_action( 'admin_notices', 'wpsc_jetpack_boost_notice' );
+}
```
I might be missing context about where is this file included, but shouldn't this be the other way around - I mean the `$_GET` check inside `wpsc_jetpack_boost_notice`?

donnchawp replied:
It's loaded by wp-cache.php, so it's loaded by the frontend and backend, but I've done it this way before, to stop loading the function unnecessarily when the stop condition was simple. Your argument for doing the other way would be to keep the function self contained?

pyronaur replied:
Mostly that. It just seems like a micro-optimization and I'm not sure I see the value in doing that. I've always leaned towards keeping hooks self-contained.

donnchawp replied:
Fair enough. I moved the checks inside. Makes sense.

pyronaur commented on projects/plugins/super-cache/js/admin.js:
```diff
@@ -9,9 +9,7 @@ jQuery( document ).ready( () => {
-	const link = jQuery( '.wpsc-install-action-button' );
-	const label = link.find( 'label' );
-	const spinner = link.find( '.spinner' );
+	let link, label, spinner, errorMessage, originalText;
```
This looks like an anti-pattern to me. I'm not sure I understand why these have to be in the upper scope. If it's to share selectors, you could place them in a funciton.

donnchawp replied:
Thanks, I wasn't happy with having them out there either, but I wanted to get this checked in yesterday evening. A class or a function is the next step.

pyronaur replied:
I tinkered with this a bit, here's untested code - I'm not sure if it works, but this is the direction in which I was thinking:


```js
/**
 * Handle the buttons for the Boost migration.
 * @param $ - jQuery
 */
( $ => {
	$( document ).ready( function () {
		// Don't run on versions of WordPress too old for the block editor and the translation methods it brings.
		// All the install / activate options are plain links with meaningful destinations anyway.
		if ( ! window.wp || ! window.wp.i18n ) {
			return;
		}

		const { __, sprintf } = window.wp.i18n;
		const ajaxurl = window.ajaxurl;
		const wpscAdmin = window.wpscAdmin;

		const setupBoostButton = $target => {
			const $label = $target.find( 'label' );
			const $spinner = $target.find( '.spinner' );
			const $errorMessage = $target.prev( '.wpsc-boost-migration-error' );
			const source = $target.attr( 'data-source' );
			const originalText = $label.text();

			// Helper function to show an error.
			const showError = err => {
				resetBoostBannerButton();

				$errorMessage
					.text(
						err ||
							__( 'An error occurred while trying to activate Jetpack Boost', 'wp-super-cache' )
					)
					.show();
			};
			// Helper function to show Boost Banner work in progress.
			const showBusy = action => {
				link.attr( 'disabled', true );
				$label.text( action );
				$spinner.addClass( 'is-active' ).show();
			};

			// Helper function to reset Boost Banner button.
			const reset = () => {
				link.attr( 'disabled', false );
				$label.text( originalText );
				$spinner.removeClass( 'is-active' ).hide();
			};

			// Activate Jetpack Boost.
			const activateBoost = () => {
				showBusy( __( 'Activating…', 'wp-super-cache' ) );

				$.post( ajaxurl, {
					action: 'wpsc_activate_boost',
					_ajax_nonce: wpscAdmin.boostActivateNonce,
					source: source,
				} )
					.done( response => {
						if ( response.success ) {
							$label.text( 'Success! Sending you to Jetpack Boost...' );
							$spinner.hide();
							window.location.href = 'admin.php?page=jetpack-boost';
						} else {
							showError( response.data );
						}
					} )
					.fail( response => {
						showError(
							sprintf(
								/* translators: %d is an HTTP error code */
								__( 'Failed to activate Jetpack Boost: HTTP %d error received', 'wp-super-cache' ),
								response.status
							)
						);
					} );
			};

			const installBoost = () => {
				showBusy( __( 'Installing…', 'wp-super-cache' ) );
				$.post( ajaxurl, {
					action: 'wpsc_install_plugin',
					_ajax_nonce: wpscAdmin.boostInstallNonce,
					slug: 'jetpack-boost',
				} )
					.done( response => {
						if ( response.success ) {
							boostActivation.activateBoost( source );
						} else {
							showError( response.data );
						}
					} )
					.fail( response => {
						boostActivation.showError(
							sprintf(
								/* translators: %d is an HTTP error code */
								__( 'Failed to install Jetpack Boost: HTTP %d error received', 'wp-super-cache' ),
								response.status
							)
						);
					} );
			};

			return {
				installBoost,
				activateBoost,
				showError,
				showBusy,
				reset,
			};
		};

		// One-click install for Boost.
		$( '.wpsc-install-boost-button' ).on( 'click', event => {
			event.preventDefault();
			const boostActivation = setupBoostButton( $( event.currentTarget ) );
			boostActivation.installBoost();
		} );

		// Handle activate button click.
		$( '.wpsc-activate-boost-button' ).on( 'click', event => {
			event.preventDefault();
			const boostActivation = setupBoostButton( $( event.currentTarget ) );
			boostActivation.activateBoost();
		} );

		// Dismiss Boost banner.
		$( '.wpsc-boost-dismiss' ).on( 'click', () => {
			$( '.wpsc-boost-banner' ).fadeOut( 'slow' );
			$.post( ajaxurl, {
				action: 'wpsc-hide-boost-banner',
				nonce: wpscAdmin.boostDismissNonce,
			} );
		} );

		// Dismiss admin notice
		$( '.boost-notice' ).on( 'click', '.notice-dismiss', event => {
			event.preventDefault();
			$.post( ajaxurl, {
				action: 'wpsc_dismiss_boost_notice',
				_ajax_nonce: wpscAdmin.boostNoticeDismissNonce,
			} );
		} );
	} );
} )( jQuery );

```

donnchawp committed:
Place JS in a function to avoid polluting global variables

donnchawp committed:
Add a MINIMUM_BOOST_VERSION cosntant for that Boost version

donnchawp committed:
Fixes for static analysis checks

donnchawp committed:
Updated phan baseline

dilirity commented on projects/plugins/super-cache/js/admin.js:
```diff
@@ -1,4 +1,7 @@
-jQuery( document ).ready( () => {
+/**
```
I think the way it was handled before is better. The callback just needs to be an unnamed function, not an arrow function.

Instead of

```js
jQuery( document ).ready( () => {
```

use

```js
jQuery( document ).ready( function () {
```

That way, `handleMigrationButtons` won't be global.

donnchawp replied:
Thanks!

dilirity replied:
@donnchawp I misled you when I said that it's better to use an unnamed function vs an arrow function. My intention was to avoid global scope for the `handleMigrationButtons` function.

Both get the job done. If there's a function inside an arrow function, it would still get scoped to the arrow function.

dilirity reviewed (dismissed):
One thing I noticed. I haven't done a complete review.

donnchawp committed:
Put it in an anon function

donnchawp committed:
Update the button handling code.

Yes, that's a terrible commit message, but it's a huge overhaul of the
JavaScript, thank you Nauris.

donnchawp committed:
Remove this class, it's not needed any more

donnchawp committed:
Dismiss the admin notice if Jetpack Boost is activated by migration

donnchawp committed:
Move the supercache settings page check into the notice.

pyronaur reviewed (approved):
:shipit:
